js/scorm/wrapper.js

```diff
@@ -940,6 +936,15 @@ class ScormWrapper {
     return response.join(',');
   }
 
+  setSessionTime() {
+    const endTime = new Date();
+    if (this.isSCORM2004()) {
+      this.setValue('cmi.session_time', this.convertToSCORM2004Time(endTime.getTime() - this.startTime.getTime()));
+    } else {
+      this.setValue('cmi.core.session_time', this.convertToSCORM12Time(endTime.getTime() - this.startTime.getTime()));
+    }
```

Comment:
```suggestion
    if (this.isSCORM2004()) {
      this.setValue('cmi.session_time', this.convertToSCORM2004Time(endTime.getTime() - this.startTime.getTime()));
      return;
    }
    this.setValue('cmi.core.session_time', this.convertToSCORM12Time(endTime.getTime() - this.startTime.getTime()));
```

Reply:
I would normally return early to prevent nesting, but it doesn't seem that beneficial in this case due to how the method only contains these conditions. Not bothered either way though if this is preferred in all cases.

Reply:
Yes it is preferred for the most part:
https://github.com/adaptlearning/adapt-contrib-spoor/blob/c6fc537b66a72287ac1415609f155ada7db1df67/js/scorm/wrapper.js#L272-L278

Use `if (expression) { } else {` only if there is something after the final `}`, such as:
https://github.com/adaptlearning/adapt-contrib-spoor/blob/c6fc537b66a72287ac1415609f155ada7db1df67/js/scorm/wrapper.js#L648-L657

Otherwise just return early. It's easier to see that the first if block is the end of the function, that scorm2004 is the exception and that the primary use-case is scorm1.2:
```js
    if (this.isSCORM2004()) {
      this.setValue('cmi.session_time', this.convertToSCORM2004Time(endTime.getTime() - this.startTime.getTime()));
      return;
    }
    this.setValue('cmi.core.session_time', this.convertToSCORM12Time(endTime.getTime() - this.startTime.getTime()));
```

`isValidCompletionStatus` is a good example of why return early is easier to read, where one must read 11 extra lines to realise that a `return false` would have been sufficient much earlier:
https://github.com/adaptlearning/adapt-contrib-spoor/blob/c6fc537b66a72287ac1415609f155ada7db1df67/js/scorm/wrapper.js#L749-L772



> Not bothered either way

I've commit the suggestion.

Reply:
To make `isValidCompletionStatus` more readable, it could be rewritten like this, where one doesn't have to read ahead to see the extent of the first exception:
```js
isValidCompletionStatus(status) { 
  status = status.toLowerCase(); // workaround for some LMSs (e.g. Arena) not adhering to the all-lowercase rule 
  if (this.isSCORM2004()) { 
    switch (status) { 
      case COMPLETION_STATE.UNKNOWN.asLowerCase: 
      case COMPLETION_STATE.NOTATTEMPTED.asLowerCase: 
      case COMPLETION_STATE.NOT_ATTEMPTED.asLowerCase: // mentioned in SCORM 2004 spec - mapped to 'not attempted' 
      case COMPLETION_STATE.INCOMPLETE.asLowerCase: 
      case COMPLETION_STATE.COMPLETED.asLowerCase: 
        return true; 
    } 
    return false; 
  }
  switch (status) { 
    case COMPLETION_STATE.NOTATTEMPTED.asLowerCase: 
    case COMPLETION_STATE.BROWSED.asLowerCase: 
    case COMPLETION_STATE.INCOMPLETE.asLowerCase: 
    case COMPLETION_STATE.COMPLETED.asLowerCase: 
    case SUCCESS_STATE.PASSED.asLowerCase: 
    case SUCCESS_STATE.FAILED.asLowerCase: 
      return true; 
  }
  return false;
} 
```

or this, which has much fewer keywords for the same output:

```js
isValidCompletionStatus(status) { 
  status = status.toLowerCase(); // workaround for some LMSs (e.g. Arena) not adhering to the all-lowercase rule 
  if (this.isSCORM2004()) { 
    return [
      COMPLETION_STATE.UNKNOWN.asLowerCase,
      COMPLETION_STATE.NOTATTEMPTED.asLowerCase,
      COMPLETION_STATE.NOT_ATTEMPTED.asLowerCase, // mentioned in SCORM 2004 spec - mapped to 'not attempted' 
      COMPLETION_STATE.INCOMPLETE.asLowerCase, 
      COMPLETION_STATE.COMPLETED.asLowerCase
    ].includes(status)
  }
  return [
    COMPLETION_STATE.NOTATTEMPTED.asLowerCase,
    COMPLETION_STATE.BROWSED.asLowerCase,
    COMPLETION_STATE.INCOMPLETE.asLowerCase,
    COMPLETION_STATE.COMPLETED.asLowerCase,
    SUCCESS_STATE.PASSED.asLowerCase, 
    SUCCESS_STATE.FAILED.asLowerCase
  ].includes(status)
}
```